File: hr1-applicant-app/lib/features/auth/data/repositories/profile_repository.dart
import 'dart:io';

import 'package:supabase_flutter/supabase_flutter.dart';

/// プロフィールデータリポジトリ
class ProfileRepository {
  const ProfileRepository(this._client);

  final SupabaseClient _client;

  /// プロフィールを取得
  Future<Map<String, dynamic>> getProfile(String userId) async {
    final response = await _client
        .from('profiles')
        .select()
        .eq('id', userId)
        .single();
    return response;
  }

  /// アバター画像をアップロードし、プロフィールを更新して公開URLを返す
  Future<String> uploadAvatar(File file, String extension) async {
    final userId = _client.auth.currentUser?.id;
    if (userId == null) throw Exception('ユーザーが認証されていません');

    final path = '$userId/${DateTime.now().millisecondsSinceEpoch}.$extension';

    await _client.storage
        .from('avatars')
        .upload(path, file, fileOptions: const FileOptions(upsert: true));

    final publicUrl = _client.storage.from('avatars').getPublicUrl(path);

    await _client
        .from('profiles')
        .update({'avatar_url': publicUrl})
        .eq('id', userId);

    return publicUrl;
  }

  /// プロフィールを更新
  Future<void> updateProfile({String? displayName, String? avatarUrl}) async {
    final userId = _client.auth.currentUser?.id;
    if (userId == null) throw Exception('ユーザーが認証されていません');

    final updates = <String, dynamic>{
      'updated_at': DateTime.now().toIso8601String(),
    };
    if (displayName != null) updates['display_name'] = displayName;
    if (avatarUrl != null) updates['avatar_url'] = avatarUrl;

    await _client.from('profiles').update(updates).eq('id', userId);
  }

  /// 企業を選択（user_organizations に登録）
  /// 既に登録済みの場合は無視する
  Future<void> selectOrganization({
    required String userId,
    required String organizationId,
  }) async {
    try {
      await _client.from('user_organizations').insert({
        'user_id': userId,
        'organization_id': organizationId,
      });
    } on PostgrestException catch (e) {
      // 23505 = unique_violation（既に登録済み）→ 無視
      if (e.code != '23505') rethrow;
    }
  }
}

File: hr1-employee-web/src/lib/repositories/profile-repository.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import type { Profile, Department } from "@/types/database";

export async function updateMyProfile(
  client: SupabaseClient,
  userId: string,
  fields: Partial<
    Pick<Profile, "display_name" | "name_kana" | "phone" | "position" | "birth_date" | "gender">
  >
) {
  const { error } = await client.from("profiles").update(fields).eq("id", userId);
  if (error) throw error;
}

export async function uploadAvatar(
  client: SupabaseClient,
  userId: string,
  file: File
): Promise<string> {
  const ext = file.name.split(".").pop() ?? "jpg";
  const path = `${userId}/${Date.now()}.${ext}`;
  const { error: uploadError } = await client.storage
    .from("avatars")
    .upload(path, file, { upsert: true });
  if (uploadError) throw uploadError;

  const {
    data: { publicUrl },
  } = client.storage.from("avatars").getPublicUrl(path);

  const { error: updateError } = await client
    .from("profiles")
    .update({ avatar_url: publicUrl })
    .eq("id", userId);
  if (updateError) throw updateError;

  return publicUrl;
}

export async function fetchEmployees(client: SupabaseClient, organizationId: string) {
  const { data } = await client
    .from("user_organizations")
    .select("user_id, profiles!user_organizations_user_id_fkey(id, email, display_name, position)")
    .eq("organization_id", organizationId);
  return (data ?? [])
    .map((row) => (row as unknown as { profiles: Profile }).profiles)
    .filter((p): p is Profile => p != null && p.role !== "applicant")
    .map((p) => ({
      id: p.id,
      email: p.email,
      display_name: p.display_name,
      position: p.position,
    }));
}

export async function fetchDepartments(client: SupabaseClient, organizationId: string) {
  const { data } = await client
    .from("departments")
    .select("*")
    .eq("organization_id", organizationId)
    .order("name");
  return (data ?? []) as Department[];
}
